### ISSUE TITLE
[Due for payment 2026-04-16] [Image][$1000] Improve NewDot image uploading experience, add large file support

### ISSUE THREAD

**Issue by chiragsalian** [Expensify/App#9402]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Upload an image in App

## Expected Result:
The image should be grayed (opacity 0.6 i.e., chatItemUnsentMessage) until its uploaded and once its uploaded it should not show the white thumbnail with loader
![image](https://user-images.githubusercontent.com/14356145/173163988-2ba7c6fe-f77c-4e68-bfdf-1510ad5646c3.png)
and instead continue showing the local thumbnail or show the final image immediately.

## Actual Result:
1. The local image is being uploaded and we see the local thumbnail,
![image](https://user-images.githubusercontent.com/14356145/173164047-79f28f8a-1ec7-4434-a038-787a2974492f.png)
2. The local image has finished uploading and now we attempt to render it which first shows up a white thumbnail with loader
![image](https://user-images.githubusercontent.com/14356145/173164076-f91ce915-87a2-4653-8284-d067baeae777.png)
3. The render completes and we see the uploaded image
![image](https://user-images.githubusercontent.com/14356145/173164089-9327bdb1-0825-4ed6-9e64-650c580ab4c9.png)

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platform:

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.76
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1654262165566879

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

cc @michaelhaxhiu, @cead22, @kidroca.

**Comment by parasharrajat** [comment]
Good change. I agree.

**Comment by parasharrajat** [comment]
@JosueEchandiaAsto Thanks for the proposal. But as you are a new contributor and have already been assigned two 2 tasks, you will have to wait before applying to new tasks. 

Contribution policy: https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#payment-for-contributions
> New contributors are limited to working on one job at a time, however experienced contributors may work on numerous jobs simultaneously.

**Comment by JosueEchandiaAsto** [comment]
@parasharrajat Both tasks have already been completed and have been merged. Do I need to complete any additional steps in those tasks?  ☹️

**Comment by parasharrajat** [comment]
Oh, I see. I will review your proposal shortly.

**Comment by parasharrajat** [comment]
@JosueEchandiaAsto Your proposal does not meet the requirements. 

> It would be better to keep it, because it is the loading that all the images have when a url is received and it is being loaded.

This spinner while loading the image is better to keep it.

Why should we keep the loader?

**Comment by JosueEchandiaAsto** [comment]
@parasharrajat 

This is the process:
1. You are requested to upload an image.
2. A thumbnail with opacity 0.6 is displayed.
3. Confirmation of successful upload is received and the list of chat messages is reloaded.
4. If any message is of type image, with the url we load the image.
5. A loading is displayed.
6. Loading the image hides the loading.

In step 3, when reloading the list of chat messages, the thumbnail is lost and what comes from the service is loaded. In this case the white loading is shown. while loading the url that comes from the service. Therefore, it seemed to me that it was better to leave the current post-upload experience.
Now, I can make it save the thumbnail, then validate which message it corresponds to and instead of showing the loading show the saved thumbnail. Is it what you want to do?

**Comment by parasharrajat** [comment]
> Now, I can have it save the thumbnail, then validate which message it corresponds to, and instead of showing the load, show the saved thumbnail. Is that what you want to do?

It would be best to achieve the expected results from this issue. let's see your proposal and we can analyze if it fits well.

**Comment by JosueEchandiaAsto** [comment]
@parasharrajat Of course yes, I implement it and send an update of the proposal. Thank you very much.

**Comment by cead22** [comment]
I think the ideal experience should be
- User uploads image
- The user sees the thumbnail of the local right away
- **Note** if the user is online, the thumbnail should show with full opacity. The transparency should only appear if the user isn't online imo (cc @chiragsalian)
- The image never changes, we don't show a loading image/spinner, to the user uploading the image it should look as if it worked **instantly**

**Comment by MitchExpensify** [comment]
@JosueEchandiaAsto Am I right in understanding you are considering a proposal for this change?

**Comment by JosueEchandiaAsto** [comment]
@MitchExpensify Hello, I have been preparing another proposal but it is complicated to recover the local image when the image post-upload service is called again. At the moment, I am looking at other Issues.

**Comment by MitchExpensify** [comment]
Got it! Thanks for that context.

**Comment by MitchExpensify** [comment]
Price doubled!

**Comment by deetergp** [comment]
Not overdue. Switching back to weekly since we are awaiting proposals.

**Comment by MitchExpensify** [comment]
Doubled!

**Comment by marktoman** [comment]
# Proposal

The uploaded image goes through two phases. First, it is shown from the local URL, then, once it’s uploaded, from the server URL. This causes flickering and the loading icon. Flickering is also caused by resizing the image from the default size when re-rendering.

### `ImageWithSizeCalculation`

First, let’s add a new `localUrl` property so we can distinguish in which phase we are at the image component level. It takes precedence over the url prop.

[https://github.com/Expensify/App/blob/773b93f296da193d030a4859b425bb9f74d366f7/src/components/ImageWithSizeCalculation.js#L107](https://github.com/Expensify/App/blob/773b93f296da193d030a4859b425bb9f74d366f7/src/components/ImageWithSizeCalculation.js#L107)

```diff
+ const uri = this.props.localUrl || this.props.url;
```

Remove `contain` mode, which causes resizing and does not change the appearance of the image because it is fully contained in the component given that its size is calculated from the image itself. Additionally, we set the `defaultSource` to the same value on the web, otherwise, the RN component flickers. There is an issue related to Avatar flickering where the contributor describes this bug extensively.

[https://github.com/Expensify/App/blob/773b93f296da193d030a4859b425bb9f74d366f7/src/components/ImageWithSizeCalculation.js#L120](https://github.com/Expensify/App/blob/773b93f296da193d030a4859b425bb9f74d366f7/src/components/ImageWithSizeCalculation.js#L120)

```diff
- source={{uri: this.props.url}}
- resizeMode="contain"
+ source={{uri}}
+ defaultSource={Platform.select({web: {uri}, default: undefined})}
```

Only show the upload icon if we are not uploading the image

[https://github.com/Expensify/App/blob/773b93f296da193d030a4859b425bb9f74d366f7/src/components/ImageWithSizeCalculation.js#L125](https://github.com/Expensify/App/blob/773b93f296da193d030a4859b425bb9f74d366f7/src/components/ImageWithSizeCalculation.js#L125)

```diff
- {this.state.isLoading && (
+ {this.state.isLoading && !this.props.localUrl && (
```

Remove the default values which cause flickering. This change does not affect the image in any way because the size is calculated.

### `ThumbnailImage`

[https://github.com/Expensify/App/blob/4b4db2365cd444d5c1f75ab7489f34b4bf60bcc5/src/components/ThumbnailImage.js#L33-L34](https://github.com/Expensify/App/blob/4b4db2365cd444d5c1f75ab7489f34b4bf60bcc5/src/components/ThumbnailImage.js#L33-L34)

```diff
- imageWidth: 200,
- imageHeight: 200,
+ imageWidth: undefined,
+ imageHeight: undefined,
```

Pass the prop through.

[https://github.com/Expensify/App/blob/4b4db2365cd444d5c1f75ab7489f34b4bf60bcc5/src/components/ThumbnailImage.js#L103-L106](https://github.com/Expensify/App/blob/4b4db2365cd444d5c1f75ab7489f34b4bf60bcc5/src/components/ThumbnailImage.js#L103-L106)

```diff
+ localUrl={this.props.localUrl}
```

Now we have to make sure that once the image is rendered in Phase 2, it is shown from the local URL instead of the remote URL, which results in no image to show until it’s downloaded. Any subsequent re-render will be fetched automatically from the remote URL.

We do that by temporarily holding the original filename through the phases, then we delete the reference.

The image tag in Phase 2 is generated from HTML received from the server. The uploading process is asynchronous and the original file name is the only way how to correlate both phases.

### `ImageRenderer`

We add a simple object that refreshes when the app loads.

[https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L9](https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L9)

```diff
+ const currentUploads = {};
```

Here, `imagePreviewModalDisabled` indicates that we are in Phase 1. If that is the case, normalize the image name because the server does it too. Set `localUrl` to the current URL. If that is not the case, we are in Phase 2, so fetch the cached local URL, assign it to `localUrl` , and delete it from the cache. [https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L50](https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L50)

```diff
+ const normFileName = originalFileName.replaceAll(/[\s-._]/g, '');
+ let localUrl = null;
+ if (imagePreviewModalDisabled) {
+     currentUploads[normFileName] = previewSource;
+     localUrl = previewSource;
+ }
+ else if (normFileName in currentUploads) {
+     localUrl = currentUploads[normFileName];
+     delete currentUploads[normFileName]
+ }
```

Assign the value to the new prop.

[https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L52](https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L52)

```diff
+ localUrl={localUrl}
```

Pass it through.

[https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L71](https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L71)

```diff
+ localUrl={localUrl}
```

In Phase 2, the server attaches the original filename to the `data-name` attribute. In Phase 1, we render the tag ourselves. We have to attach it the same way in order to be resolved to the `originalFileName` variable used above.

### `ReportActionItemFragment`

[https://github.com/Expensify/App/blob/c9eea6c71e68abb088474cecf3f62ac4b870ae08/src/pages/home/report/ReportActionItemFragment.js#L77](https://github.com/Expensify/App/blob/c9eea6c71e68abb088474cecf3f62ac4b870ae08/src/pages/home/report/ReportActionItemFragment.js#L77)

```diff
- <RenderHTML html={`<comment><img src="${props.attachmentInfo.source}" data-expensify-preview-modal-disabled="true"/></comment>`} />
+ <RenderHTML html={`<comment><img src="${props.attachmentInfo.source}" data-name="${props.attachmentInfo.name}" data-expensify-preview-modal-disabled="true"/></comment>`} />
```

**Comment by kidroca** [comment]
> Remove the default values which cause flickering. This change does not affect the image in any way because the size is calculated.

Would that work for remotely served images? 

If the size is calculated from the image it has to first be fetched, what will happen in the meantime (e.g. slow connection)
It might work instantly for local images but will have delay for served content

**Comment by marktoman** [comment]
We can set it conditionally instead. If the image is rendered when uploading (including Phase 2), keep it undefined, otherwise set the original default values.

**Comment by deetergp** [comment]
I'm back from my day off and will be reviewing the proposal shortly…

**Comment by deetergp** [comment]
@marktoman Your proposal seems pretty well reasoned but I'd love to see it in action, especially with a slow/throttled internet connection.

**Comment by deetergp** [comment]
Thanks for the screen captures @marktoman. I like your proposal. If @parasharrajat agrees, then let's go with it!

**Comment by deetergp** [comment]
@parasharrajat Any thoughts to add?

**Comment by marktoman** [comment]
> Or Do you think we can preserve the localstate? If we can, this could have been achieved using URLs only.

We cannot because each comment image rendered on the screen creates a new `ImageRenderer` instance. When uploading, we get two independent instances originating from two places:
1. When we start uploading, in which case we control the instance ([here](https://github.com/Expensify/App/blob/544821fc5254e975d81d10e32e7a472c5af29ad6/src/pages/home/report/ReportActionItemFragment.js#L77)). This instance is disposed of when the 2. point happens.
2. When the server sends us the HTML ([here](https://github.com/Expensify/App/blob/544821fc5254e975d81d10e32e7a472c5af29ad6/src/pages/home/report/ReportActionItemFragment.js#L103-L104)). Our uploaded image is now part of the comments with no relation to the uploading event.

> Is it the best place to store this?

I believe it is because `ImageRenderer` is the top-level component in this context and already has the uploading-related logic (code that uses `imagePreviewModalDisabled`). Splitting this high-level logic and drilling part of it down to the low-level components has no benefit and goes against the separation of concerns principle. We should pass down only the resolved value that is required by the low-level component.

**Comment by parasharrajat** [comment]
I got it. So what if we keep the same RenderHTML instance and just update the HTML. Will it still create a new instance of ImageRenderer? I guess it will. What do you think?

**Comment by marktoman** [comment]
The event in 2. point happens regardless. If we preserve the `RenderHTML` instance created in 1. to 2., we still have to somehow decide that what we receive in 2. is the same image as 1. and ignore it, otherwise we end up with two representations of the same image on the screen. (Deciding at this level would require additional parsing, so it is better to take advantage of the regular parsing process and decide at the `ImageRenderer` level.)

**Comment by parasharrajat** [comment]
Ok, we can move forward with this implementation for now. A suggestion would be move out the currentuploads logic to another lib and use it inside IMGrenderer. 

@marktoman's proposal looks good to me.

cc: @deetergp 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by cead22** [comment]
> You can see the image flickering for a split second, but I'm guessing that's a delay from loading image from local storage

Is there a way to prevent the flickering showed in the video for native [here](https://github.com/Expensify/App/issues/9402#issuecomment-1179045149)? I see it didn't happen on the native video [here](https://github.com/Expensify/App/issues/9402#issuecomment-1179045149), was it fixed or is there a different reason this happened on one video but not the other?

**Comment by marktoman** [comment]
The quote is from a different proposal. Both videos reference my proposal, which does not cause flickering and the videos do not flicker.

**Comment by cead22** [comment]
Awesome, thanks for clarifying

**Comment by deetergp** [comment]
Thanks for the discussion, everyone. @MitchExpensify let's hire @marktoman to move forward with his proposal 👍

**Comment by cead22** [comment]
Any updates on this?

**Comment by deetergp** [comment]
Not overdue, waiting on a PR. Bumping it down to `Weekly`.

**Comment by MitchExpensify** [comment]
How is the PR coming along @marktoman ?

**Comment by MitchExpensify** [comment]
How is the PR coming along @marktoman ?

**Comment by marktoman** [comment]
I am experiencing inconsistent behavior on ios. I can push it as a draft PR by Monday, but I'll need more time after that to find a way.

**Comment by MitchExpensify** [comment]
Any update on this one @marktoman?

**Comment by marktoman** [comment]
Let me give it another try this week to finally decide if this problem makes my solution unworkable.

**Comment by MitchExpensify** [comment]
Awesome, thanks @marktoman!

**Comment by michaelhaxhiu** [comment]
I think this issue has dragged on for a long period of time. It would be really great to push forward on this as soon as we can because it's adversely affecting a core part of the New Expensify experience.

Also can someone confirm if we are going to **kill this text** as a part of improving the experience? I think it's super old school and unnecessary 🔪 

<img width="577" alt="image" src="https://user-images.githubusercontent.com/13206896/188173927-730abf26-bd55-4aeb-8d92-c58d49b79cef.png">

**Comment by marktoman** [comment]
@parasharrajat Do I still have this week to conclude or suddenly not?

**Comment by marktoman** [comment]
@parasharrajat You cannot open it to new proposals after I got confirmed that I have this week to improve the proposal that you have approved. Nobody self-respecting can work like that.

**Comment by MitchExpensify** [comment]
> Do I still have this week to conclude or suddenly not?

We'll definitely honor that timeline @marktoman. Lets see if we can come up with a valid solution in the next 3/4 days and go from there

**Comment by MitchExpensify** [comment]
@deetergp Mind assigning @marktoman to the job? Thank you

**Comment by marktoman** [comment]
@parasharrajat Thanks for clarifying. I was hoping you don't mean it right away, but it could be read that way. I'll give it one last try over the weekend and post a conclusion.

**Comment by MitchExpensify** [comment]
Sounds good, thanks @marktoman!

**Comment by marktoman** [comment]
The issue, which affects all name-based proposals, is that if a new file has the same name as a previously uploaded one (e.g., invoice.png), the old file is rendered the same as the newly uploaded one.

My recent effort was to find a way to distinguish them. I am convinced that this is not possible without server changes. The change would be to generate an id for each uploaded file on the client and temporarily store it on the server. The ids can be deleted after x hours because we don't care about old files. The client would use a library like nanoid and send the id alongside the file name. The rest is the same, except we would read the id instead of the file name when distinguishing the newly uploaded file.

**Comment by trjExpensify** [comment]
>The issue, which affects all name-based proposals, is that if a new file has the same name as a previously uploaded one (e.g., invoice.png), the old file is rendered the same as the newly uploaded one.

Ah, interesting. @marcaaron & @tgolen worked on better front-end validation for attachment uploads in this [issue](https://github.com/Expensify/App/issues/10050). I wonder if they have some insight here?

**Comment by tgolen** [comment]
OK, I am not familiar with this issue, but I took some time to read through the comments. I'm not sure I am 100% there, but I don't think we need any server-side changes for this. That sounds overly complex. When uploading an image, can we just update the image name to have the current timestamp in it like `invoice_<timestamp>.png`?

**Comment by tgolen** [comment]
but I'm not quite sure what the intention of this is. If all you're trying to do is distinguish between the localURL and the HTML hosted URL, can you search the URL for the s3 domain?

**Comment by marktoman** [comment]
I thought the name should be retained for the user, but if that is not the case, then yes, it can be changed to an id. The intention is to identify a local file and the uploaded remote file as the same. When the user adds a file, it is rendered from a local path, then once uploaded, it comes back from the server and is rendered again from the URL.

**Comment by tgolen** [comment]
I think it'll be OK to add a timestamp to the name. It at least keeps the original name intact so it doesn't look totally foreign to the user, but there is no requirement for keeping the filename 100% the same as the original (that I am aware of).

**Comment by marktoman** [comment]
Basically, the solution is to prevent Phase 2 (ie, the file coming from the server once uploaded) from re-rendering the image.

> I think if we have a "local path" for a file that we would just keep using that forever as long as it exists (and only then access the file from the server)

This is the behavior.

**Comment by marcaaron** [comment]
>the solution is to prevent Phase 2 (ie, the file coming from the server once uploaded) from re-rendering the image

@marktoman can you maybe re-state what the proposal is at a high level without referring to phases? I missed what you are proposing.

**Comment by marktoman** [comment]
@marcaaron This is what happens currently without my solution implemented, which are 2 isolated events:
1. The user adds an image. The image is rendered from the local URL. At the same time, the image starts uploading to the server.
2. Once uploaded, the server sends back an IMG element with a remote URL. The client renders the IMG as any other HTML coming from the server.

My solution correlates both events:

1. When the user adds a file, store both the file name (which can be an id as discussed) and the local path in a temporary map on the client.
2. Once the server sends the IMG tag, if its name attribute matches a value stored previously (point 1), use the local path instead of the remote URL coming with the IMG tag.

**Comment by tgolen** [comment]
Are there any standard best practices for react native apps when it comes to loading/caching images?

I am concerned about using a temporary in-memory cache for this because it is only going to live as long as the memory of the app (which can be all over the place).

**Comment by kidroca** [comment]
I think the current issue is solely focused on fixing the upload experience and misses an opportunity to address a broader problem 

Instead of a temporary map that only servers to cover attachment uploads, we can do the following 

#### Let the `Image` component handle rendering
1. We're trying to render a report action containing an image
2. The `Image` component first checks if we have a local copy in our cache dir and it instantly loads it (this convers offline usage as well)
3. Otherwise the `Image` component has to download (and then cache) content to display it

#### How the above fixes the Upload experience:
1. When we upload an attachment we immediately add it to cache 
2. There's no "content flash" because the `Image` component is not trying to download the image from s3 in order to render it

Having this generic logic added to the Image component should
- reduce networking (load from cache instead of making request)
- increase image loading speed (loading local content vs s3)
- images work offline (as long as they've been seen or uploaded by the current user)
- current ticket is also addressed

**Comment by tgolen** [comment]
Yes, thanks! That's much more inline with what I think a true fix would be
for this and sounds like what I would expect to happen.

**Comment by marktoman** [comment]
> 2. There's no "content flash" because the Image component is not trying to download the image from s3 in order to render it

There are two challenges:
1. Loading from the remote URL (the loading icon)
2. Replacing the local image instance with a new image instance coming from the server (content flash)

Your overview covers 1., but not 2. The 2. point was more time-consuming and it is different for each platform. You can take it from my solution.

**Comment by tgolen** [comment]
> Replacing the local image instance with a new image instance coming from the server (content flash)

My understanding of what @kidroca was saying is that this would no longer happen. We shouldn't ever be switching out a locally cached image with an instance coming from the server. If it's fetched from the server, it's because there is no local image.

**Comment by marktoman** [comment]
When the user adds an image, a locally created element is added at the end of the chat. Once the server receives the image content, it sends a new chat HTML that contains the new image. I am not sure how is the locally created element preservable in the stream of new messages.

**Comment by kidroca** [comment]
> When the user adds an image, a locally created element is added at the end of the chat. Once the server receives the image content, it sends a new chat HTML that contains the new image. I am not sure how is the locally created element preservable in the stream of new messages.

Yeah, during upload we have an optimistic report action with attachment that's rendered using raw data
After upload it gets replaced by a report action containing `html` rendered by the HTML Parser
##### Sample uploaded attachment
![Expensify Screenshot 2022-09-07 143046](https://user-images.githubusercontent.com/12156624/188902249-fd688aa4-ec2c-41ec-a4c8-1fa063f2f584.png)

I guess the HTML Parser should parse the `img` tag to our special react-native image component, then it can use the `data-name` attribute/prop for cache purposes - if a file by that name exists in cache storage use it, otherwise fetch & cache


@parasharrajat 
> We have to figure out how we want to store the images locally in a cross-platform way.

I think we already figured this out here: https://github.com/Expensify/App/issues/3867 
- ios/android - Onyx saves paths to local files
- web/desktop - files are saved to IndexedDB 

they can be feeded as image source

**Comment by deetergp** [comment]
There's been quite a lot of discussion on this and I'm not sure where we landed. Did we decide that backend changes are required and that this is held up on internal changes, or are we able to forge ahead with updates to the front end?

**Comment by deetergp** [comment]
Bump @parasharrajat @marktoman?

**Comment by parasharrajat** [comment]
Even, I am not sure where we are landed on this. It seems that we discussed not to do https://github.com/Expensify/App/issues/9402#issuecomment-1172569896 in the above comments. 

If we are looking to do https://github.com/Expensify/App/issues/9402#issuecomment-1237932342, then it would be good to set a framework to save any attachments instead of just images. Then https://github.com/Expensify/App/issues/6527 comes into the picture before this and thus it should be tracked with https://github.com/Expensify/App/issues/10894

**Comment by Beamanator** [comment]
So quick update: We're putting many (pretty much all) Image-related feature request and improvements on a list (see [this tracking issue](https://github.com/Expensify/App/issues/10894)) and we're putting them on HOLD until we design a complete solution for these types of issues. Therefore I'm going to put this issue on HOLD for now.

I love all the ideas so far from @marktoman and @kidroca , rest assured we'll work to compensate everyone fairly who has contributed useful ideas / solutions 👍

**Comment by deetergp** [comment]
Not overdue, just on hold.

**Comment by deetergp** [comment]
Still on hold

**Comment by mallenexpensify** [comment]
@MitchExpensify Snaggin' from you cuz I got the tracking issue assigned to me
[Tracking] [Polish] Image-related feature request and improvements #10894

**Comment by mallenexpensify** [comment]
on hold

**Comment by trjExpensify** [comment]
This issue is held, Melv.

**Comment by chiragsalian** [comment]
I'm not following why this issue is on HOLD for a tracking issue - https://github.com/Expensify/App/issues/10894. Maybe that was a mistake and its meant to be on HOLD for something else. Anyone knows?

**Comment by trjExpensify** [comment]
@Beamanator commented above [here](https://github.com/Expensify/App/issues/9402#issuecomment-1270431340). More specifically, it's under the umbrella of image rendering/flashing and this pool of issues is held on implementing better caching.  

<img width="952" alt="image" src="https://user-images.githubusercontent.com/16232057/202040008-11021299-5835-41ac-9d64-ebef23f28f6d.png">

Now that we are out of planning and have specific issues for the imagine caching, we can update the title to cite #10792 & #12603 specifically.

**Comment by trjExpensify** [comment]
👋 The native caching PR has merged, so we're down to the one for web/desktop. Updating the title to reflect that we're held only on #12603 at this point.

**Comment by trjExpensify** [comment]
No change, Melv. Still on hold for the caching PRs

**Comment by trjExpensify** [comment]
No change, caching PRs still ongoing.

**Comment by deetergp** [comment]
Still on hold

**Comment by trjExpensify** [comment]
No change, that PR is still ongoing.

**Comment by deetergp** [comment]
Still on hold

**Comment by deetergp** [comment]
Still on hold

**Comment by trjExpensify** [comment]
Good news, the native side of caching merged today. Web is soon to follow hopefully!

**Comment by deetergp** [comment]
It looks like https://github.com/Expensify/App/issues/12603 is still actively being worked on, so this is still on hold.

**Comment by JmillsExpensify** [comment]
@trjExpensify on that linked issue, are we waiting on the upstream fix or have we already forked?

**Comment by trjExpensify** [comment]
We've forked, and we're waiting on this [App PR](https://github.com/Expensify/App/pull/13036) to test it all. Update is [here](https://github.com/Expensify/App/issues/12603#issuecomment-1410057871).

**Comment by deetergp** [comment]
Still on hold.

**Comment by trjExpensify** [comment]
Same, Melv.

**Comment by trjExpensify** [comment]
Same, held on cashing that's held on a CORS issue.

**Comment by deetergp** [comment]
Still on hold

**Comment by JmillsExpensify** [comment]
Same same.

**Comment by deetergp** [comment]
Still on hold

**Comment by deetergp** [comment]
Still on hold.

**Comment by trjExpensify** [comment]
Given that caching is held on a CORS issue and that's now held on the server migration, I'm dropping this to monthly until we get closer to a resolution on those two dependencies.

**Comment by deetergp** [comment]
Still on hold.

**Comment by deetergp** [comment]
Still holding

**Comment by trjExpensify** [comment]
No change from the chain of dependencies [here](https://github.com/Expensify/App/issues/9402#issuecomment-1531116199).

**Comment by trjExpensify** [comment]
No change from the chain of dependencies [here](https://github.com/Expensify/App/issues/9402#issuecomment-1531116199).

**Comment by trjExpensify** [comment]
No change from the chain of dependencies https://github.com/Expensify/App/issues/9402#issuecomment-1531116199.

**Comment by trjExpensify** [comment]
Getting closer! CORS issue resolved, web image caching PR in development. :)

**Comment by trjExpensify** [comment]
In light of #32703 being fixed, the [new image caching PR](https://github.com/Expensify/App/pull/34505) is in review.

**Comment by trjExpensify** [comment]
The image caching PR was [reverted again](https://github.com/Expensify/App/pull/35636) on Friday, so we're back to figuring that out in #12603.

**Comment by quinthar** [comment]
What is the next step here: waiting for #12603 to ship and then what?  Who is doing it?

**Comment by trjExpensify** [comment]
@kidroca is [this still](https://github.com/Expensify/App/issues/9402#issuecomment-1237932342) accurate, so once #12603 is out the door there won't actually be any follow-up work necessary here?

**Comment by kidroca** [comment]
> @kidroca is [this still](https://github.com/Expensify/App/issues/9402#issuecomment-1237932342) accurate, so once #12603 is out the door there won't actually be any follow-up work necessary here?

While #12603 moves attachment authentication to headers and promotes better caching it won't fully solve the upload experience process, because we don't have control over native browser cache.
What I'm talking about in [comment you refer to](https://github.com/Expensify/App/issues/9402#issuecomment-1237932342) is that we can add small alterations and leverage the Cache API (which is different from native browser cache) 
- Once we upload an attachment we instantly add it to cache
- When we try to load remote attachments we first check cache
- in case we have cache the image component renders instantly
- out of the box result: the image can be displayed while uploading

#### Potential drawbacks 
- A similar solution would have to be applied for native iOS/Android 
- We have to manage cache ourselves - this can be a thin cache layer, or a more complicated one depending on needs, but the least we'll have to monitor quota and cleanup older files 

#### Alternative solution that is simpler
- use the local image during uploads
- drawback: it only work for the current browser session
   - if we're offline and the image didn't upload the next time we open a chat we won't see the image

**Comment by trjExpensify** [comment]
Cool, going to solicit a few opinions to weigh in on that.

**Comment by trjExpensify** [comment]
Hey @kidroca, I circled this internally and we'd prefer to explore the path that adds a more robust caching solution than simply relying on local storage -- which is viable, but only for the current browser session, after which we could very well lose your attachments.

@roryabraham had some parting thoughts:

- The Cache API sounds appropriate for this use case, but I'm pretty unclear on what the native equivalent would be. If it's very much file-driven, maybe the API could be implemented more consistently across platforms if we leverage OPFS on web instead.
- Leveraging service workers could be a way to improve performance of our caching layer, such that the caching doesn't happen on the main thread
- I think it would be a good idea to design a caching solution holistically with more use-cases in mind (i.e: caching html documents, fonts, and the JS bundle as well, in addition to images). We don't need to include this in the V1 but we should at least think about it. 

Is this a project you would be interested in fleshing out and taking the lead on?

**Comment by kidroca** [comment]
Hey @trjExpensify thanks for considering me for this project. 

I'm aligned with @roryabraham points and ready to present and explore our options 

To ensure I'm on the right track, could you clarify what "fleshing out" entails? Are we looking for a detailed proposal or a preliminary outline of potential approaches? 

Looking forward for your guidance and eager to get started.

**Comment by roryabraham** [comment]
@kidroca I think a high-level overview of your proposed approach, formatted as a **Problem** / **Solution** would be the best way to move this forward. In fact, just aligning on a clear problem statement seems like a great first step.

Then, once you spend some time with the problem and do some initial investigation, you can host a "predesign" in slack, where you present the problem and the potential solutions. Keeping this high-level (and in plain English, no code) is a best practice.

Then, once we've aligned on the best high-level approach, you can produce a detailed implementation plan for the solution we've agreed upon, again using mostly plain English rather than code.

Finally, once we've aligned on the problem and the detailed implementation plan for the solution, we can begin implementation in earnest and review the code.

Does that plan of action make sense?

1. Problem statement
1. Predesign high level solutions in #expensify-open-source
1. Predesign detailed implementation of our preferred solution, either in #expensify-open-source or in this GH issue
1. Implement

I look forward to seeing what you produce. Don't hesitate to reach out if you get stuck or have questions. Thanks!

**Comment by quinthar** [comment]
Put a larger plan here: https://expensify.slack.com/archives/C01GTK53T8Q/p1708849229702429?thread_ts=1707270017.261459&cid=C01GTK53T8Q

**Comment by quinthar** [comment]
Ok, I think the plan is:
1. @kidroca [HIGH: Add support for inline image Markdown](https://github.com/Expensify/App/issues/37246) -- This allows us to insert images into the actual HTML itself, which is a precursor to everything.
2. Maybe @tgolen - Once we have the above, [[HOLD ON 37246] HIGH: Combine image attachment with the unsent composed message](https://github.com/Expensify/App/issues/35977)?
3. Then we do this issue, which is really evolving into "add large object uploading".

**Comment by trjExpensify** [comment]
Nice! For (3) do you want to use this issue, or start afresh with a new "clean" one if it's evolving into a larger plan for large object uploading?

**Comment by kidroca** [comment]
> 3\. Then we do this issue, which is really evolving into "add large object uploading".

Let's create a new ticket for large file uploads. The current ticket aims to enhance the upload interface, as shown in the screenshot. We plan to replace the blank rectangle with a grayed-out image preview during uploads. Addressing large uploads is a separate issue and won't fix this visual gap.

![current-upload-xp](https://user-images.githubusercontent.com/14356145/173163988-2ba7c6fe-f77c-4e68-bfdf-1510ad5646c3.png)

**Comment by trjExpensify** [comment]
Cool, I'll keep this one open then for the attachment thumbnail on upload. :)

**Comment by trjExpensify** [comment]
@kidroca with #12603 out the way, are we holding this issue on https://github.com/Expensify/App/issues/35977 now or can we move forward with it as part of #3 from [here](https://github.com/Expensify/App/issues/9402#issuecomment-1966375805)?

**Comment by kidroca** [comment]
> @kidroca, now that #12603 is resolved, does #35977 hold us back, or are we clear to proceed with it as outlined in #3 [here](https://github.com/Expensify/App/issues/9402#issuecomment-1966375805)?

I believe we're not blocked by #35977. Although I'm currently assigned to several tickets, my next focus will be enhancing the upload experience, and I'll be tracking this effort under the current ticket.

For additional context, here's [a relevant discussion from Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1711392756174079?thread_ts=1707270017.261459&cid=C01GTK53T8Q).

> - auth tokens are moved from query params to http headers, this stabilizes the URL and improves caching
> - I don't think anything is actionable, we need to plan and account for all platforms
> 
> I find the current strategy suitable, but it has to be revised a bit, in short it is:
> 1. present a slightly transparent thumbnail of the local file during upload (instead of the text "Uploading attachment...")
> 2. keep using the local file until close or switch to a different report
> 3. the next time we open the report - load attachments from their remote source, this would also cache them
> 4. if the attachment is not yet uploaded we still present the local attachment as the optimistic action points to that
> The caching layer idea, where uploads are added directly to local cache can be added on top, but this step requires platform specific implementations and managing cache ourselves

**Comment by trjExpensify** [comment]
Perfect, excited for it! If you do end up needing backend support for this task, give us a shout. 👍

**Comment by trjExpensify** [comment]
How's it going, @kidroca?

**Comment by deetergp** [comment]
Any update @kidroca?

**Comment by kidroca** [comment]
Hey team, apologies for the radio silence - I've been tied up with some unexpected priorities. I'll be out of office until the end of the week. I'll be focusing on this ticket and making some key enhancements to the upload process.

**Comment by trjExpensify** [comment]
Nice, look forward to it!

**Comment by trjExpensify** [comment]
👋 @kidroca what's the plan for this week?

**Comment by kidroca** [comment]
**Status Update:**

I've been planning the implementation of the upload enhancements. However, I've had to postpone this ticket due to more critical issues that needed my attention.

**Current Plan:**

- Use conditional rendering to display an `<Upload>` component instead of raw text.
- Append local `src` metadata to the report action to render thumbnails without flashes.
- For images, display a thumbnail; for documents or other files, display a generic thumbnail.
- Consider directly loading from the backend, allowing a visible transition for user confirmation.

I'll resume work on this ticket this week.

**Comment by deetergp** [comment]
Hey @kidroca. Any progress to report?

**Comment by kidroca** [comment]
Hi, I'm currently tied up with two critical issues that are consuming all of my available time. You can find more details here: https://expensify.slack.com/archives/C05LX9D6E07/p1716402826532549

**Comment by trjExpensify** [comment]
👋 @kidroca where is this one in the queue now?

**Comment by kidroca** [comment]
I've resumed work on this issue and aim to post a PR by end of week

**Comment by kidroca** [comment]
Status update:

I'm working on this, but haven't posted a PR yet

**Comment by kidroca** [comment]
### Status Update

I've successfully displayed the local image during upload, but to prevent a content flash, we need a backend update.

1. When the user uploads a file, we work with the local file and create an optimistic report action with it.
2. Later, when the action is swapped by the backend-created action, we lack a reliable way to match the remote image address to our local file.
3. This mismatch causes the image component to load the remote URL, resulting in a brief flash.

To avoid this flash, we can generate a unique attachment ID on New Dot via `NumberUtils.rand64()` (similar to how report action IDs are generated) and send it as part of the attachment comment payload (under `attachmentInfo`). The backend can then use it as a data attribute on the `<img>` component added to the report action (`data-attachment-id`). Alternatively, we can send the desired `<img>` HTML as part of the attachment comment. I've done that, but it seems the backend is currently stripping any extra `<img>` attributes I've added.

This approach would allow us to resolve the attachment from local data:
- For upload cases, this would avoid the content flash.
- It would save network bandwidth by resolving user-owned attachments from local data.
- It would ensure instant loading for other cases.
- If the user deleted the content, it would be loaded from the remote source.

Here's a brief demonstration of the current implementation and the content flash:

https://github.com/Expensify/App/assets/12156624/0edb3a52-7d6b-44d8-a85e-5c9174f8275c

**Comment by trjExpensify** [comment]
Did we bring this out to Slack yet? Personally, I wouldn't be against splitting this out into two steps. The first being to ship what you've got to display the thumbnail instead of `Uploading attachment...` which is a good improvement in itself, and the second step to address the content flash. CC: @Expensify/design @JmillsExpensify as well.

**Comment by dannymcclain** [comment]
Yeah I'm not against splitting it out either. I agree that displaying the thumbnail instead of `Uploading attachment...` is an improvement. And while the content flash is annoying, it's not the end of the world. So taking care of that in a follow up would be ok with me.

**Comment by shawnborton** [comment]
Yup, I agree with that sentiment as well.

**Comment by kidroca** [comment]
Ok, I'll post an update that adds image / video preview during attachment upload

I'm tidying up edge cases and testing across all platforms

**Comment by trjExpensify** [comment]
👋 How you getting on @kidroca?

**Comment by kidroca** [comment]
### Status Update

Hi team,

I'm back from OOO and will be posting a PR for this feature today. I'm experiencing some issues building the Android app, but I'll post the PR regardless and work on resolving the issue later.

I've encountered a code change that introduced a new element `<uploading-attachment>` (details here: [PR 42516](https://github.com/Expensify/App/pull/42516/files#diff-577fcc5f3a5e4930916ff310e61c38edfa093792fe3ab55b9a30d2aecd8811dbR3497)), which affects my changes. 
I've posted a comment requesting more information on the related issue
- https://github.com/Expensify/App/issues/42297#issuecomment-2210595399

**Comment by kidroca** [comment]
I'm having trouble viewing images on local web, I have the same problem on latest `main` and I'm not sure what the problem is. If I go to new.expensify.com the images do appear


This is the only thing similar I've discovered on slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1719897451272009?thread_ts=1714117518.672049&cid=C01GTK53T8Q

Images just appear as forever loading, not just the ones I send now, but all the past attachments that I have in my chats

**Comment by kidroca** [comment]
### Status Update

I was able to verify my PR on all platforms and it's ready for review:
- https://github.com/Expensify/App/pull/44889

**Comment by kidroca** [comment]
### Status Update

✅ The PR has been reviewed.

Several bugs were discovered during the review, some of which were pre-existing. We have decided to address these in follow-up PRs. I will take responsibility for these follow-ups and report progress as we resolve each item. The issues related to image upload are as follows:

- [ ] **Bug: Preview closes automatically and chat is redirected to another chat**
    - This will be addressed by leveraging local content when available instead of switching to the remote source which causes the original item to disappear and auto-close the carrousel.
- [ ] **Bug: No preview while sending a video offline**
    - When offline, attempting to send a video results in a "You appear to be offline" screen when trying to open the carousel.
- [ ] **❓Bug: No preview for PDFs until the message is replaced with server response**
    - The backend generates a thumbnail image for document previews. Currently, a generic download link is displayed for documents until they are uploaded and replaced by the backend.
        <details>
        <summary>This is how PDF uploads look like. Do we need to take further action here?</summary>
        
        https://github.com/user-attachments/assets/7b0f82f0-fcde-4837-a7c3-b4aacb943389
        </details>
- [ ] **Bug: Video upload does not display the play button in the middle of the thumbnail**
    - Details: [Issue Comment](https://github.com/Expensify/App/pull/44889#issuecomment-2231190097)
- [ ] **Bug: Quick flash or blink when the image transitions from the optimistic version to the actual uploaded version**
    - This will be addressed by leveraging local content when available instead of loading the remote source.

I will post my plan for the next follow-up PR in a new comment and keep this list updated as we check off each item.

**Comment by kidroca** [comment]
I'd like to elaborate on the next steps for enhancing content management by utilizing local data effectively.

### Implementing Data Attributes for Efficient Local Content Management

As we've discussed, the primary enhancement I propose is utilizing local content when available. To map remote sources to local content effectively, we need to introduce an `attachmentID` generated by the client and included in requests. This identifier has been a topic in our recent Slack discussions, particularly in our latest thread which can be found here: [Slack Conversation](https://expensify.slack.com/archives/C01GTK53T8Q/p1721929571364809?thread_ts=1707270017.261459&cid=C01GTK53T8Q).

The implementation would involve adding a `data-attachment-id` attribute to HTML elements like images, videos, and anchors. This attribute would act both as an identifier and as a potential final name for the attachment. However, this requires a backend modification to the `AddAttachment` and `AddTextAndAttachment` actions, as currently, they do not retain any data attributes from the HTML I'm sending.

My proposal is to include this `data-attachment-id` on the returned `<img>`, `<video>`, and `<a/>` tags for content uploaded by users. Since my last PR, we've transitioned from sending placeholders like "Uploading image..." to sending actual HTML markup for attachments. The issue now is that any additional html attributes I send are not retained, the backend seems to process replace the html and sends img/video/anchors with the default data attributes and values.

Once implemented, this identifier would allow us to cache local paths to attachments. This means that:
- Attachments will render immediately using the local path, preventing the flash that occurs when local sources swap with the remote sources.
- When loading a chat, attachments with locally available paths would load instantly, while those without local copies would still fetch from the backend as usual.
- This approach will also address the "Preview gets closed" bug, as the source of an attachment during and after upload would consistently be the same local path.
- Local paths would be cached inside Onyx during upload. They will be keyed by the unique attachment id and we can store some additional file metadata like dimensions

Regarding other issues, I'll update my list of follow up work. If I find any of these issues easy to fix, I will prioritize them as my next follow-up. Otherwise, my focus will remain on implementing the main enhancement discussed here. 

Meanwhile, discussing and setting the backend changes in motion would be beneficial.

**Comment by kidroca** [comment]
Opened a follow-up PR to fix the console error when uploading pdfs
- Issue: https://github.com/Expensify/App/issues/46615#top
- PR: https://github.com/Expensify/App/pull/47184

The PR is marked as draft as I haven't yet uploaded videos and tested on all platforms

**Comment by trjExpensify** [comment]
Just removed the payment holds for melv. 👍

**Comment by kidroca** [comment]
The follow-up PR is ready for review, but has an eslint error that I need to fix
- https://github.com/Expensify/App/pull/47184

The rest of the changes are blocked as they need the backend update described here: https://github.com/Expensify/App/issues/9402#issuecomment-2266099020

**Comment by trjExpensify** [comment]
>The rest of the changes are blocked as they need the backend update described here: https://github.com/Expensify/App/issues/9402#issuecomment-2266099020

@deetergp are you working on those?

**Comment by deetergp** [comment]
> @deetergp are you working on those?

I will look into these later today.

**Comment by deetergp** [comment]
I was not able to get into this this week but will jump in it first thing when I'm back on Tuesday.

**Comment by deetergp** [comment]
Still have not had a chance to look into this deeper. Hoping to do so tomorrow.

**Comment by deetergp** [comment]
In response to [this comment](https://cooking.nytimes.com/recipes/1020527-chicken-vesuvio) calling for back-end changes to allow `data-attribute-id`, I believe I figured out where we are stripping it from `AddAttachment` & `AddTextAndAttachment` and how to add it back. @kidroca do you have a WIP PR that I could use to test these back end changes?

**Comment by deetergp** [comment]
Rebump @kidroca

**Comment by deetergp** [comment]
> @deetergp Looks like you like cooking. This link takes to a recipe.

Haha, yes I was cooking. Dish turned out to be pretty tasty too!

This is the comment from @kidroca I had intended meant to link to https://github.com/Expensify/App/issues/9402#issuecomment-2266099020 since Github has decided to bury it.

**Comment by deetergp** [comment]
I'm going to move this to Weekly till we hear back from @kidroca.

**Comment by deetergp** [comment]
Still no news from @kidroca. You still out there?

**Comment by deetergp** [comment]
Still no word from @kidroca. Do we need to open this back up to other contributors?

**Comment by trjExpensify** [comment]
Do we have a clear picture of what we need done in the front-end PR for that? In the meantime, I've reached out. 👍

**Comment by kidroca** [comment]
Hi team, I apologize for my extended absence and lack of communication. I'm back now and reviewing the current state of the ticket. I'll provide an update within 24 hours, after I've caught up with the latest changes

**Comment by kidroca** [comment]
## Status Update
I've synchronized my PR with latest changes from `main`.
- https://github.com/Expensify/App/pull/47184#top

- I've tried including `data-attachment-id` to the markup we send when creating attachments, but re-inspecting the code I've realised this markup is only being used in the optimistic data.
- We can generate `attachmentID` and send it as part of the parameters for `API.write` (details below).
- next: I'll re-test the fix in the PR and add screenshots so it's ready for review.

We could do the following:
1. When we create the attachment report action (on the client)
2. We generate an attachmentID via `NumberUtils.rand64()`
3. We add a parameter like `parameters.attachmentData` to the "AddAttachment" API, where we store attachment related information
4. When the backend creates the action report action and markup it can read the attachment-id from `parameters.attachmentData.attachmentID` 

Or instead of `parameters.attachmentData` we can introduce, `parameters.attachments` and move any attachment information like `file` there. This way when we support multiple attachments they can be listed in `parameters.attachments` as `[{file: ..., attachmentID: ...}, {...}, ...]`

---

My original understanding was that "AddAttachment" API worked with the html markup generated on the client, so the original proposal was to add `data-attachment-id` to the markup generated here:
https://github.com/Expensify/App/blob/30d508b1e9c2b8598965ffc2e722f830a0a44cfb/src/libs/ReportUtils.ts#L4234-L4248

But inspecting the Network call I see no markup is being sent, or at least the `<img>` part is being omitted, double checked the code at `Report.addActions` and it seems that indeed we send only `reportCommentText`

**Comment by deetergp** [comment]
The issue here is that on the back end, we get the file information from PHP's built-in `$_FILES` global which has a pre-defined set of parameters. I'm a little uncertain about how we'd get the attachmentID to track with the files 🤔 
<img width="1656" alt="Screenshot 2024-11-07 at 10 39 40 AM" src="https://github.com/user-attachments/assets/53319264-bed8-46ea-9791-e5961bb22eb6">

It's possible we could pass a separate `parameters.attachments` array then loop through the `$_FILES` array and try to associate the  additional metadata passed in the `attachments` array with it.

**Comment by deetergp** [comment]
Just so I am crystal clear on what we need on the back end here: The end result we are looking for here is that that we will add `data-attachment-id=<the ID passed from NewDot>` to the image tag in the HTML we save to the `message` field on the ADDCOMMENT report action. Is that correct?

**Comment by kidroca** [comment]
> Just so I am crystal clear on what we need on the back end here: The end result we are looking for here is that that we will add `data-attachment-id=<the ID passed from NewDot>` to the image tag in the HTML we save to the `message` field on the `ADDCOMMENT` report action. Is that correct?

Yes

> It's possible we could pass a separate `parameters.attachments` array then loop through the `$_FILES` array and try to associate the additional metadata passed in the `attachments` array with it.

Yes, both these lists should have the same order when sent

I originally thought that we send file metadata as JSON, but it seems we only send binary data for the file, so any metadata like the attachment id should be added under a separate key

Since we currently support only a single attachment let's add a key like `attachmentMeta` or `attachmentData` to the payload for `AddAttachment` and `AddTextAndAttachment`

Or if we like to keep it super simple we can just add the key `attachmentID` to these actions

**Comment by kidroca** [comment]
> Just so I am crystal clear on what we need on the back end here: The end result we are looking for here is that that we will add `data-attachment-id=<the ID passed from NewDot>` to the image tag in the HTML we save to the `message` field on the `ADDCOMMENT` report action. Is that correct?

Actually we need to add this to all attachments not just `<img>`, but `<video>` and `<a>` for attachments as well

**Comment by deetergp** [comment]
> Or if we like to keep it super simple we can just add the key attachmentID to these actions

Let's keep it simple for now. I have a back end PR updated and out for review.

**Comment by deetergp** [comment]
The [PR to add this](https://github.com/Expensify/Web-Expensify/pull/43547) is now on staging.

**Comment by trjExpensify** [comment]
Great stuff, that's now on prod! @kidroca can you give us an idea of when you can progress the front-end PR? Thanks!

**Comment by deetergp** [comment]
Bump @kidroca

**Comment by mallenexpensify** [comment]
Randomly ended up here, bumped to `Daily` and made @kidroca the owner to get 👀.  I'll ping him too

**Comment by deetergp** [comment]
I removed @kidroca & @marktoman & reapplied the `Help Wanted` label

**Comment by trjExpensify** [comment]
👋 yo yo! Can you make sure to link to Slack threads referenced to follow the breadcrumbs? Also, there's a ton of history with this issue. It would be helpful to post a fresh comment to summarise what we're asking someone to take on to follow Scott's backend PR. Thanks!

**Comment by trjExpensify** [comment]
@parasharrajat what's your ETA on the proposal review?

**Comment by trjExpensify** [comment]
>I am collecting the old work around image to be able to suggest next step on the proposal. I should have it by tomorrow.

@parasharrajat do you have an ETA on this?

**Comment by deetergp** [comment]
Looks like we're making some progress

**Comment by deetergp** [comment]
@NJ-2020 What is the API command you are calling when you are trying to pass the `attachmentID`?

**Comment by deetergp** [comment]
I wasn't able to get to this today, but I'll try to get you a good answer tomorrow morning.

**Comment by trjExpensify** [comment]
Any luck, Scott?

**Comment by deetergp** [comment]
My apologies gang, I've been pretty tied up on other projects. I'll try to get you an answer later today.

**Comment by deetergp** [comment]
This will be top of my todos tomorrow morning!

**Comment by deetergp** [comment]
Okay I think I understand why you weren't getting an `attachmentID` @NJ-2020 

> I cannot test this because when I passed the attachmentID param to the BE, the BE does not return img tag with data-attachment-id attribute, I've tried using attachment-id, files: [{file, attachmentID: rand64()}] but still all didn't work

It's counterintuitive, to be sure, but NewDot currently only supports single image uploads, so despite attachments/files being an array, you can only add one. The `attachmentID` is a separate parameter that you send alongside `files`. Does that make sense?

**Comment by deetergp** [comment]
Sorry folks! Yes I've been a little busy, but I was able to spend some time on this today and think I understand why you weren't getting an attachment ID — that param was never added to our WAF so it was always 0 no matter what. After adding it to the WAF, I've confirmed that calling `AddAttachment` and passing it an `attachmentID` param does create a reportAction that includes `data-attachment-id`. That HTML does not appear to be used when it gets displayed in NewDot.

![](https://private-user-images.githubusercontent.com/766197/423728710-5a1fa0c5-61af-4949-9438-b16cd357943c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyNTkyMDEsIm5iZiI6MTc0MjI1ODkwMSwicGF0aCI6Ii83NjYxOTcvNDIzNzI4NzEwLTVhMWZhMGM1LTYxYWYtNDk0OS05NDM4LWIxNmNkMzU3OTQzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOFQwMDQ4MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMmEyOTdjNDQ3YjQ4ZDI5ZjQ3YmFiZmY1YTE0Y2NhNGMwOWI5NmNjZTAyMWYzNmQzNmNjNzc3MDJlY2VhYmE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JgLLxT3zdwuGfXIIRieQcppJ21SU_-Oqb633aXlROYk)
![](https://private-user-images.githubusercontent.com/766197/423728763-c4d64d0d-deda-41e7-8023-89de216fbd7a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyNTkyMDEsIm5iZiI6MTc0MjI1ODkwMSwicGF0aCI6Ii83NjYxOTcvNDIzNzI4NzYzLWM0ZDY0ZDBkLWRlZGEtNDFlNy04MDIzLTg5ZGUyMTZmYmQ3YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOFQwMDQ4MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZWRmMWMwNGVmZmEyMDg4Mjg4NTAwMDhkNjU4YjAwOWE4MWZjYmMzYjNlNTI3ZTgwY2U0ZWFlMjA3ODVlNjI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Xo68pyVnPk8BM-tegsFDi9sw2siCzj3o0rsizu8ZCas)
![](https://private-user-images.githubusercontent.com/766197/423728501-57e61c13-a303-4e57-912c-1f5042e1a6fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyNTkyMDEsIm5iZiI6MTc0MjI1ODkwMSwicGF0aCI6Ii83NjYxOTcvNDIzNzI4NTAxLTU3ZTYxYzEzLWEzMDMtNGU1Ny05MTJjLTFmNTA0MmUxYTZmYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOFQwMDQ4MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZmQzNmNhZTE0OTY4YWY5OWI5YjY2NDFmMDQyMTFiOTA3Y2FkZWJiNDdjN2ZlZTA4OGE0MjMzMTk3Mjc5Yjk2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sA_ZozOMgVCgFj6mHpK5J-5nA5AzIvA0j6tfeuR6NIo)
<img width="1824" alt="Image" src="https://github.com/user-attachments/assets/37285c64-8f05-4e29-9cba-756dea223b30" />

**Comment by deetergp** [comment]
@NJ-2020 Sorry, the code to fix it had to be reviewed & deployed before you could test it. It has been approved and is was deployed to staging about an hour ago, so you ought to be able to test it there.

**Comment by deetergp** [comment]
Nice! Thanks for letting us know @NJ-2020.

**Comment by deetergp** [comment]
Thanks for the update @NJ-2020 👍

**Comment by mallenexpensify** [comment]
@parasharrajat , 👀 above plz

**Comment by NJ-2020** [comment]
I am still working on it, will provide an update today

**Comment by NJ-2020** [comment]
Hi 👋 Sharing an update today:

- Enable Cache API support to electron
- Implement Cache API into libs folder
- Initialize Cache API when opening the app

But there's some challenges that I am facing right now, which is, when retrieving the value from the Cache storage, it returns a Promise value which we can't use directly into image `src`, so to resolve that we need to return the original type

I am going to investigate and provide a fix solution

Thanks!

**Comment by parasharrajat** [comment]
@NJ-2020 Could you please share the steps where will be either setting or getting the attachment source from storage or to storage?

**Comment by NJ-2020** [comment]
@parasharrajat Sure, here it's https://github.com/Expensify/App/commit/fa58efa6abca48f24d485154ecc4e17171bf558d

> But there's some challenges that I am facing right now, which is, when retrieving the value from the Cache storage, it returns a Promise value which we can't use directly into image src, so to resolve that we need to return the original type

I think we can make something like i.e useOnyx but for cache api since it's returning promise value, wdyt?

> Enable Cache API support to electron

Regarding Cache API support for Electron, I reverted those changes due to an error. 
Actually it works well on desktop when I try running cache api on dev tools, but I think we should verify the desktop compatibility first since there were previous issues with Cache API on desktop (see [here](https://github.com/Expensify/App/pull/59188) and [here](https://github.com/Expensify/App/issues/59184))."

**Comment by NJ-2020** [comment]
And also I am still investigating the react-fast-pdf ~ react-pdf render issue

**Comment by parasharrajat** [comment]
Let's keep the scope of this issue minimal. We can break it down into multiple PRs if needed. 

1. First PR, add support for web and native to save attachments. Use unit tests to verify that it works. 
2. Second, implement this to chat attachments. 

Any other issues? 

1. List them down here... we will target them one by one.

**Comment by parasharrajat** [comment]
> Could you please share the steps where will be either setting or getting the attachment source from storage or to storage?

I mean, in the app, which components require getting or saving to the cache? Use cases. 


We should create a hook to fetch images for us. full abstraction from the complexities of inner workings. You simply pass the URL or attachment id, and you get back the cached attachment. 


Let's speed this one up. I want us to move faster here and close this issue asap.

**Comment by parasharrajat** [comment]
Once you are ready, please create a summary again, mentioning each implementation detail.

**Comment by parasharrajat** [comment]
For desktop, Let's bring this back to slack. We might have to use native file system for this as well if cache does not work like we are doing on mobile.

**Comment by NJ-2020** [comment]
> I mean, in the app, which components require getting or saving to the cache? Use cases.

For saving attachment it's inside the `addActions` function right here, but we might also need to handle for text images i.e like `![](https://images.unsplash.com/...)`
https://github.com/NJ-2020/Expensify/blob/e60f38e8553c3b432d006ac1932cc18d070916ce/src/libs/actions/Report.ts#L668

For getting/displaying attachments, the relevant files are:
1. [ImageRenderer](https://github.com/NJ-2020/Expensify/blob/e60f38e8553c3b432d006ac1932cc18d070916ce/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.tsx#L73)
2. [VideoRenderer](https://github.com/NJ-2020/Expensify/blob/e60f38e8553c3b432d006ac1932cc18d070916ce/src/components/HTMLEngineProvider/HTMLRenderers/VideoRenderer.tsx#L31)
3. [AnchorRenderer](https://github.com/NJ-2020/Expensify/blob/e60f38e8553c3b432d006ac1932cc18d070916ce/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.tsx#L120)

> For desktop, Let's bring this back to slack. We might have to use native file system for this as well if cache does not work like we are doing on mobile.

Okay

> Once you are ready, please create a summary again, mentioning each implementation detail.

Sure, here it's:

> First PR, add support for web and native to save attachments. Use unit tests to verify that it works.

So for the First PR I will add support for web and native to save attachments only and add unit tests to make sure everything is working

> Second, implement this to chat attachments.

For the second one, I will implement it and render the attachment into report chat action i.e like Image/Video/AnchorRender and make sure everything works smoothly without any issues

Thank you

**Comment by NJ-2020** [comment]
> We should create a hook to fetch images for us. full abstraction from the complexities of inner workings. You simply pass the URL or attachment id, and you get back the cached attachment.

Sure, but I think if the attachment doesn't have any attachment id, I think we can use the report action id + the attachment index instead of using the attachment url since it's better and easier to manage

So it will look something like this `{reportActionID}_{index}` {index} << will be the number of list of images in 1 reportAction i.e carousel

**Comment by parasharrajat** [comment]
> Sure, but I think if the attachment doesn't have any attachment id, I think we can use the report action id + the attachment index instead of using the attachment url since it's better and easier to manage

How will you update the local cache when the resource changes from backend or async?

**Comment by NJ-2020** [comment]
I think even the resource is changed from the BE I think it will still work because we're using the reportActionID of the report action chat and + the {index} which is mapped out to the list of img/video/anchor inside the report action

**Comment by parasharrajat** [comment]
When the resource changes, the reportaction id and index will not change, how will you identify that resource is changed? 

We can't depend on index as actions are lazy loaded and paginated, we can't say which is the index of a action. I believe we need attachment ID. Why do you think that it will be missing?

**Comment by NJ-2020** [comment]
> how will you identify that resource is changed?

Everytime the user upload any attachment in type of text/file, we will store the first remote source into Onyx, if there's any changes coming from BE, everytime we render the attachment we will compare the report action img src attribute with the Onyx remoteSource property, if it's different, then we will re-cache the attachment with the new source

**Comment by NJ-2020** [comment]
I am still working on the first PR

**Comment by NJ-2020** [comment]
Update:

For the first PR, it's almost complete, currently I am working on the take photo and take from gallery, it's because after we take a photo/ or pick any attachment from gallery, the uri file path is created inside the caches directory
https://github.com/react-native-image-picker/react-native-image-picker/blob/d2bc97ae73ef3e0e37c235dceb76f7bce9fa5f58/android/src/main/java/com/imagepicker/Utils.java#L58-L63

So if we restart our app/device the attachment file will gone because we clear it every time the app is unmounted 
https://github.com/Expensify/App/blob/81b1b49300048b3cc3a7d1d73e01b9be9873fbd2/src/libs/actions/Session/clearCache/index.native.ts#L1-L7

To fix this we need to create the file inside the documents app directory i.e `com.expensify.dev/files/.....` by using the `keepLocalCopy` from `@react-native-documents/picker` since react-native-image-picker doesn't provide any options to override the copy file path dir

and retest to make sure everything is working

**Comment by NJ-2020** [comment]
[PR](https://github.com/Expensify/App/pull/63723) (1/2) ready for review!

cc: @parasharrajat

**Comment by deetergp** [comment]
Thanks for the update @NJ-2020

**Comment by mallenexpensify** [comment]
@parasharrajat it's been weeks since you've shared a substantive update in the PR (from what I saw with a quick peek),  please prioritize. 

> I think we have already invested a lot of time implementing this. We should plan to ship it in parts as quickly as possible.

This _makes sense_ to me too.  Headups @deetergp since the PR (or PRs) are likely heading your way soon. 

(also... just realized I'm not BZ here, downside of being subscribed to emails :ohnothing:, unsubbing now.)

**Comment by parasharrajat** [comment]
We will let you know soon @NJ-2020. We are currently working on the P/S statement for this issue based on the product expectations that align with the mentioned future project bcz this issue is very old and does not have clear directions.

**Comment by NJ-2020** [comment]
Thanks

**Comment by Julesssss** [comment]
Raised [here](https://expensify.slack.com/archives/C05LX9D6E07/p1770727758572739?thread_ts=1749063014.932359&cid=C05LX9D6E07) as a potential issue for improved app performance. Might be worth prioritising this again.

**Comment by NJ-2020** [comment]
Please let me know if you need help with my [PR](https://github.com/Expensify/App/pull/65321) (continue ->)

**FYI:** We stop working on the PR because we're going to confirm the expected result first before moving forward, more info here: https://github.com/Expensify/App/pull/65321#issuecomment-3357449235, https://github.com/Expensify/App/pull/65321#issuecomment-3449340413, https://github.com/Expensify/App/pull/65321#issuecomment-3425674606 & https://github.com/Expensify/App/pull/65321#issuecomment-3465885425.

Thanks

**Comment by NJ-2020** [comment]
> Let's create the P/S and a planned task before we continue `@Julesssss`.

Yes, I think it would better to have P/S + planned task before continuing our work, and please let me know if I should create new PR, because the current [PR](https://github.com/Expensify/App/pull/65321) is a bit old & has lot of conversations.

**Comment by NJ-2020** [comment]
Based on [this](https://github.com/Expensify/App/pull/65321#issuecomment-3885983384), it seems we need immediate of caching receipts & avatars (specifically JPG images only)

<hr />
Can someone please help clarify these questions?:
<br /><br/>

- **For this [PR](https://github.com/Expensify/App/pull/65321), do we want to focus on receipt & avatars caching only, and do follow-up later, or both of them (image attachment & [receipt/avatars])?** Since we want to focus on JPG/JPEG images only
- **Onyx attachment data structure**, currently after we cached an attachment, it will be look like this in our Onyx DB:
```tsx
"attachment_123456789": {
    /** Attachment ID of the attachment */
    attachmentID: "123456789",

    /** Local source of the attachment */
    source: "file://...",

    /** Remote source url of the attachment */
    remoteSource: "https://staging.expensify.com/chat-attachments/7589660131684871223/w_1e571abcb6f1ea1403b112d9e401e3d43d52c7e3.png"
}
```
For receipts & avatars, do we want to use the same structure and only change the ID to `accountID` or || `receiptID`? or maybe something like this:
```tsx
image_{id}: {
  type: 'attachment' | 'avatar' | 'receipt', // either one of these
  source: "....",
  remoteSource: "...",
  id: "...."
}
```
- **Just wanted to make sure**:
   - For tracking/identifying image id, for avatar, do we want use `personalDetail.accountID` for avatar? and receipt maybe `transactionID` since `transaction.receipt.receiptID` only available after BE response
   - Currently we store cached attachments to `DocumentDirectoryPath` for native platform, are we okay with that?

Thanks

**Comment by Julesssss** [comment]
> For this https://github.com/Expensify/App/pull/65321, do we want to focus on receipt & avatars caching only, and do follow-up later, or both of them (image attachment & [receipt/avatars])? Since we want to focus on JPG/JPEG images only

Both 'd say, keeping a general solution. Though for implementation split PRs are fine if it helped speed up the implementation. 

> For receipts & avatars, do we want to use the same structure and only change the ID to accountID or || receiptID?

Good question. Lets prepare to discuss this with the wider team for their input. Would you mind sharing a few more details of the complexity and trade-offs with the general ID/type solutions. Initially I am thinking that `type` adds complexity, and there may not be much benifit from having the `image` collection grouping multiple types. But let me look further into this tomorrow.

**Comment by NJ-2020** [comment]
> Both 'd say, keeping a general solution. Though for implementation split PRs are fine if it helped speed up the implementation.

I think for this [PR](https://github.com/Expensify/App/pull/65321) we can proceed with the current solution which is attachments to speed up the progress, because for receipt & avatars we need to align on some key questions

> Good question. Lets prepare to discuss this with the wider team for their input. Would you mind sharing a few more details of the complexity and trade-offs with the general ID/type solutions.

Sure 👇

> Initially I am thinking that type adds complexity, and there may not be much benifit from having the image collection grouping multiple types. But let me look further into this tomorrow.

Yes, I think it's unnecessary, because using ID's I think it's already enough to identify the image (since each will have unique ID), maybe we can just use `attachment_{attachmentID || receiptID || avatarID}`:

```tsx
// attachments

"attachment_1201046928037904863": {
    attachmentID: "1201046928037904863",

    source: "file://...",

    remoteSource: "..."
}

// avatars

"attachment_21328118": {
    attachmentID: "21328118",

    source: "file://...",

    remoteSource: "..."
}

// receipts

"attachment_7023315770507700966": {
    attachmentID: "7023315770507700966",

    source: "file://...",

    remoteSource: "..."
}
```

We can also change `attachments_` to something like `media_{id}` or `cached_image_{id}`

**Comment by Julesssss** [comment]
I just want to make sure I understand the problem that we're solving here. It's been a while!

It seems to me now that `expo-image` is already handling the caching of images on disk now. `Avatar` explicitly sets `cachePolicy="memory-disk"`. Is that known? Or is that not fully solving the issue?

Our original problem is that our optimistic image renders a file that is different from the uploaded server image, and that the image flashes, right? And our solution to this is to continue using the local file, in place of the server image?

**Comment by Julesssss** [comment]
@NJ-2020 it seems preferable to implement once against `expo-image`, rather than rip out the CacheAPI stuff later. Yet rather than blocking on that, we could limit your changes to the custom storage, strictly for user-uploaded files? That would keep the upload flash fix and attachment_id mapping, but we would not implement the caching of remote images the users is simply viewing (not uploading). 

We can let `expo-image` handle remote images the user is viewing on native, and handle caching of web images separately once web is using `expo-image`?

**Comment by NJ-2020** [comment]
@parasharrajat Could you please start reviewing/testing this [PR](https://github.com/Expensify/App/pull/65321)? Thanks!

---

@Julesssss This [PR](https://github.com/Expensify/App/pull/65321) covers only the store/caching functionality — the offline preview mode will be handled in the next PR (2/2). Should we go ahead and include the full offline preview in this PR, or keep it for the next one?

Also, for caching allowed image file types, should we follow `expo-image`'s (list supported [file types](https://docs.expo.dev/versions/latest/sdk/image/#supported-image-formats)) or use a custom list? If custom, could you share the types you have in mind?

**Comment by Julesssss** [comment]
@NJ-2020 sounds good. I retriggered the adhoc build for you

**Comment by Julesssss** [comment]
PR was updated with some HEIC logic. We're getting close 🤞 

Once this PR is merged I'm going to resolve this issue and we'll continue the improvements in respective sub-issues.

**Comment by Julesssss** [comment]
> Hey [@Julesssss](https://github.com/Julesssss), unfortunately it's not 100% true. As I've [mentioned before](https://github.com/Expensify/App/issues/33725#issuecomment-3946554488), the `expo-image` has problems with caching if headers are provided

Ah right, yeah. We have [this](https://github.com/Expensify/App/issues/83112) issue to track the auth issues. Thanks for corrections.

**Comment by Julesssss** [comment]
Image caching PR review is awaiting review. I asked QA to perform additional adhoc testing [here](https://github.com/Expensify/App/pull/65321#issuecomment-3993659124).
